accepted/wasm-gc.md

```diff
@@ -0,0 +1,1362 @@
+
+impl TryFrom<EqRef> for I31Ref { /* ... */ }
+impl TryFrom<EqRef> for StructRef { /* ... */ }
+impl TryFrom<EqRef> for ArrayRef { /* ... */ }
+
+impl EqRef {
+    pub fn ref_type(&self) -> RefType { /* ... */ }
+    pub fn is_instance_of(&self, store: impl AsContext, ty: RefType) -> bool { /* ... */ }
+    pub fn is_null(&self) -> bool { /* ... */ }
+
+    pub fn is_i31_ref(&self) -> bool { */ ... */ }
+    pub fn as_i31_ref(&self) -> Option<I31Ref> { /* ... */ }
+    pub fn into_i31_ref(self) -> Option<I31Ref> { /* ... */ }
+
+    pub fn is_struct_ref(&self) -> bool { */ ... */ }
+    pub fn as_struct_ref(&self) -> Option<StructRef> { /* ... */ }
+    pub fn into_struct_ref(self) -> Option<StructRef> { /* ... */ }
+
+    pub fn is_array_ref(&self) -> bool { */ ... */ }
+    pub fn as_array_ref(&self) -> Option<ArrayRef> { /* ... */ }
+    pub fn into_array_ref(self) -> Option<ArrayRef> { /* ... */ }
+}
+```
+
+Note that the `into_*_ref` methods avoid reference counting operations. More one
+this below.
+
+`I31Ref` allows getting its inner 31-bit value as either a signed `i32` or
+unsigned `u32` and creation from those types as well (with a check that the
+value fits in 31 bits):
+
+```rust
+pub struct I31Ref {
+    // ...
+}
+
+impl PartialEq for I31Ref { /* ... */ }
+impl Eq for I31Ref { /* ... */ }
+
+impl I31Ref {
+    pub fn signed(x: i32) -> Option<I31Ref> { /* ... */ }
+    pub fn unsigned(x: u32) -> Option<I31Ref> { /* ... */ }
+
+    pub fn get_signed(&self) -> i32 { /* ... */ }
+    pub fn get_unsigned(&self) -> u32 { /* ... */ }
+}
+```
+
+The `StructRef` Rust type represents the Wasm super type for all `struct`s. It
+allows getting and setting fields.
+
+```rust
+pub struct StructRef {
+    // ...
+}
+
+impl PartialEq for StructRef { /* ... */ }
+impl Eq for StructRef { /* ... */ }
+
+impl StructRef {
+    pub fn null(ty: StructType) -> StructRef { /* ... */ }
+
+    pub fn new(
+        store: impl AsContextMut,
+        ty: StructType,
+        fields: impl Iterator<Item = Val>,
+    ) -> Result<StructRef> {
+        // ...
+    }
+
+    pub fn is_null(&self) -> bool { /* ... */ }
+    pub fn struct_type(&self) -> StructType { /* ... */ }
+    pub fn is_instance_of(&self, store: impl AsContext, ty: StructType) -> bool { /* ... */ }
+
+    pub fn get(&self, store: impl AsContextMut, field: u32) -> Result<Val> { /* ... */ }
+    pub fn set(&self, store: impl AsContextMut, field: u32, value: Val) -> Result<()> { /* ... */ }
+}
+```
+
+And the `ArrayRef` Rust type, representing the Wasm super type for all `array`s,
+looks very similar to `StructRef` but additionally has a `len` method:
+
+```rust
+pub struct ArrayRef {
+    // ...
+}
+
+impl PartialEq for ArrayRef { /* ... */ }
+impl Eq for ArrayRef { /* ... */ }
+
+impl ArrayRef {
+    pub fn null(ty: ArrayType) -> ArrayRef { /* ... */ }
+
+    pub fn new(
+        store: impl AsContextMut,
+        ty: ArrayType,
+        elements: impl Iterator<Item = Val>,
```

Comment:
For implementing `array.new` and `array.new_default`, it may be useful to have a way to construct `ArrayRef` from a single element value, which will be used for all fields. Or maybe constructing an iterator that returns the same value again and again is so easy that you can just rely on that.

Reply:
Great idea, thanks!